lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,880 @@
+
+pub const DELETED_FILE_NAME: &str = "id_tracker.deleted";
+pub const MAPPINGS_FILE_NAME: &str = "id_tracker.mappings";
+pub const VERSION_MAPPING_FILE_NAME: &str = "id_tracker.versions";
+
+#[derive(Debug)]
+pub struct ImmutableIdTracker {
+    path: PathBuf,
+
+    deleted: BitVec,
+    deleted_wrapper: MmapBitSliceBufferedUpdateWrapper,
+
+    internal_to_version: Vec<SeqNumberType>,
+    internal_to_version_wrapper: MmapSliceBufferedUpdateWrapper<SeqNumberType>,
+
+    mappings: PointMappings,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct PointMappings {
+    pub(crate) internal_to_external: Vec<PointIdType>,
+
+    // Having two separate maps allows us iterating only over one type at a time without having to filter.
+    pub(crate) external_to_internal_num: BTreeMap<u64, PointOffsetType>,
+    pub(crate) external_to_internal_uuid: BTreeMap<Uuid, PointOffsetType>,
+}
+
+/// Used endianness for storing PointMapping-files.
+type FileEndianess = LittleEndian;
+
+impl PointMappings {
+    const EXTERNAL_ID_NUMBER_BYTE: u8 = 0;
+    const EXTERNAL_ID_UUID_BYTE: u8 = 1;
+
+    /// Loads a `PointMappings` from the given reader. Applies an optional filter of deleted items
+    /// to prevent allocating unneeded data.
+    pub fn load<R: Read>(mut reader: R, filter: Option<&BitSlice>) -> OperationResult<Self> {
+        // Deserialize the header
+        let len = reader.read_u64::<FileEndianess>()? as usize;
+
+        let mut internal_to_external = Vec::with_capacity(len);
+        let mut external_to_internal_num: BTreeMap<u64, PointOffsetType> = BTreeMap::new();
+        let mut external_to_internal_uuid: BTreeMap<Uuid, PointOffsetType> = BTreeMap::new();
+
+        // Deserialize the list entries
+        for i in 0..len {
+            let (internal_id, external_id) = Self::read_entry(&mut reader)?;
+
+            // Need to push this regardless of point deletion as the vecs index represents the internal id
+            // which would become wrong if we leave out entries.
+            internal_to_external.push(external_id);
+
+            let deleted = filter
+                .as_ref()
+                .and_then(|deleted| deleted.get(i).as_deref().copied())
+                .unwrap_or_default();
+
+            if deleted {
+                continue;
+            }
+
+            match external_id {
+                ExtendedPointId::NumId(num) => {
+                    external_to_internal_num.insert(num, internal_id);
+                }
+                ExtendedPointId::Uuid(uuid) => {
+                    external_to_internal_uuid.insert(uuid, internal_id);
+                }
+            }
+        }
+
+        // Check that the file has ben fully read.
+        #[cfg(debug_assertions)] // Only for dev builds
+        {
+            let mut buf = vec![];
+            let read_bytes = reader.read_to_end(&mut buf).unwrap();
+            assert_eq!(buf.len(), 0);
+            assert_eq!(read_bytes, 0);
+        }
```

Comment:
Maybe replace `read_to_end` with [`Read::bytes`] and [`Iterator::count`]? This, most likely, won't ever be a problem in practice, but the fix is trivial.

```suggestion
        // Check that the file has been fully read.
        debug_assert_eq!(
            reader.bytes().map(Result::unwrap).count(),
            0,
        );
```

[`Read::bytes`]: https://doc.rust-lang.org/std/io/trait.Read.html#method.bytes
[`Iterator::count`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count